### PR TITLE
[vpj] dict compression for empty push to the hybrid store

### PR DESCRIPTION
**Pull request by gaojieliu**
For many real-time only stores, it is quite common to schedule an empty push to deprecate the stale entries.
This code change adds the dict compression support for the empty push to the hybrid stores.

In the meantime, this code change also fixes a bug in the existing KIF dict compression support, and it will decompress the payload in the source topic if the compression is enabled.

## How was this PR tested?
CI

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java [hunk 1]
```diff
@@ -151,12 +195,28 @@ public synchronized byte[] trainDict() {
             mapperValue = recordReader.createValue();
           }
           while (recordReader.next(mapperKey, mapperValue)) {
-            byte[] value = ByteUtils.extractByteArray(mapperValue.value);
-            currentFilledSize += value.length;
+            /**
+             * We can only decompress full compressed value here.
+             * If the source version is using {@link CompressionStrategy.NO_OP}, the value will be passed to the dict directly.
+             * If the source version is using other compression strategies, since we can only decompress the full value here, and the
+             * chunked values will be skipped.
```
Comment (FelixGV) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:202]: That is fine for now, but in some edge case it might lead to suboptimal compressibility. For example, if chunking kicks in only when a subset of the fields (which contain very large content) are populated, then it would be ideal to train the dictionary on those big fields. It is an edge case so no need to worry about it for now, but maybe having a comment to that effect may be useful.

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.system.SystemProducer;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class TestEmptyPush {
```
Comment (FelixGV) [inline internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java:42]: Is this new test a long-running one? If it is, consider adding it to one of the faster running letter buckets in `internal/venice-test-common/build.gradle`

**Review by FelixGV — COMMENTED** [review]
LGTM! Just a couple of minor comments. Thanks a lot Gaojie. Looking forward to seeing this capability in action!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java (hunk 1 above)
Comment (gaojieliu) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:202]: Done

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java (hunk 2 above)
Comment (gaojieliu) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java:42]: The test will run for less than 1 min and I added it to bucket `H`, which is one of the fastest buckets today.

**Review by m-nagarajan — CHANGES_REQUESTED** [review]
Thanks a lot Gaojie. LGTM, left a comment on the test cases.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 3]
```diff
@@ -1737,6 +1734,46 @@ private Optional<ByteBuffer> getCompressionDictionary() throws VeniceException {
 
       return Optional.ofNullable(compressionDictionary);
     } else {
+      /**
+       * Special handling for an empty push to a hybrid store.
```
[the remark this answers is not in the document]
Comment (gaojieliu) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1738]: Done

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        .build();
+    venice = ServiceFactory.getVeniceCluster(options);
+  }
+
+  @AfterClass(alwaysRun = true)
+  public void tearDown() {
+    Utils.closeQuietlyWithErrorLogged(venice);
+  }
+
+  @Test
```
[the remark this answers is not in the document]
Comment (gaojieliu) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java]: Added a test timeout: 2min

I was trying to avoid adding too many integration tests to extend the CI running time.
I think the modified tests should be good enough to validate these scenarios:
1. First empty push with dict compression enabled to a hybrid store without any records in RT.
2. Writing some records to this hybrid store and validate them.
3. Second empty push with Gzip enabled and validate the records.
4. Third empty push with dict compression enabled, and the dict built this time will be different from 1), which would build some dummy dict.
5. Fourth empty push with dict compression enabled, and the dict built this time should be the same as the 3rd empty push.

[the remark this answers is not in the document]
Comment (FelixGV) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java]: I don't think we should shy away from comprehensive testing... the tests run in parallel now so it's not as much a concern as before (as long as there isn't a single test class that runs super long...).

[the remark this answers is not in the document]
Comment (m-nagarajan) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestEmptyPush.java]: Thanks Gaojie. This test covers all the scenarios I was checking about.

**Review by m-nagarajan — APPROVED** [review]
LGTM, Thanks Gaojie.
